providers/keycloak.go

```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
+			_, ok := roles.([]interface{})
+			if !ok {
+				return nil, errors.New("error parsing realm roles from claims")
+			}
+			for _, r := range roles.([]interface{}) {
+				roleList = append(roleList, fmt.Sprintf("%s", r))
+			}
+		}
+	}
+
+	if clientRoles, found := claims["resource_access"].(map[string]interface{}); found {
+		for name, list := range clientRoles {
+			scopes, ok := list.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("error parsing client roles from claims")
+			}
+			if roles, found := scopes["roles"]; found {
+				for _, r := range roles.([]interface{}) {
+					roleList = append(roleList, fmt.Sprintf("%s:%s", name, r))
+				}
+			}
+		}
+	}
+
+	return roleList, nil
+}
+
+func (p *KeycloakProvider) getUserInfo(ctx context.Context, s *sessions.SessionState) (*simplejson.Json, error) {
+	userInfo, err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
 		UnmarshalJSON()
 	if err != nil {
 		logger.Errorf("failed making request %s", err)
-		return "", err
+		return nil, fmt.Errorf("error getting user info: %v", err)
 	}
+	return userInfo, nil
+}
 
-	if p.Group != "" {
-		var groups, err = json.Get("groups").Array()
+func getEmailAddress(userInfo *simplejson.Json) (string, error) {
+	var email, err = userInfo.Get("email").String()
+	if err != nil {
+		return "", fmt.Errorf("error getting email address: %v", err)
+	}
+	return email, nil
+}
+
+func getGroups(userInfo *simplejson.Json) ([]string, error) {
+	if g, ok := userInfo.CheckGet("groups"); ok {
+		groups, err := g.StringArray()
 		if err != nil {
-			logger.Printf("groups not found %s", err)
-			return "", err
+			return []string{}, fmt.Errorf("error getting groups: %v", err)
 		}
+		return groups, nil
+	}
+	return []string{}, nil
+}
 
+func getRoles(s *sessions.SessionState) ([]string, error) {
+	claims := make(map[string]interface{})
+	// Decode JWT token without verifying the signature
+	token, err := jwt.ParseSigned(s.AccessToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse token: %v", err)
+	}
+	// Parse claims
+	if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		logger.Printf("failed to parse claims: %s", err)
+	}
+	roles, err := extractRolesFromClaims(claims)
+	if err != nil {
+		logger.Printf("failed to extract roles: %s", err)
+	}
+	return roles, nil
+}
+
+func (p *KeycloakProvider) isAuthorizedGroupMember(s *sessions.SessionState) (bool, error) {
+	if p.Group != "" {
 		var found = false
-		for i := range groups {
-			if groups[i].(string) == p.Group {
+		for i := range s.Groups {
+			if s.Groups[i] == p.Group {
 				found = true
 				break
 			}
 		}
-
 		if !found {
-			logger.Printf("group not found, access denied")
-			return "", nil
+			return false, errors.New("group not found, access denied")
+		}
+	}
+	return true, nil
+}
+
+func (p *KeycloakProvider) hasRequiredRole(s *sessions.SessionState) (bool, error) {
```

Comment:
This is also unneeded in favor of the new `SetAllowedGroups` & `Authorize` functionality

Reply:
Got it, looking at that now that actually makes this much simpler and cleaner. One part I'd like to discuss here though, I've noticed that this would effectively result in an ANY behaviour between roles and groups which may not fit here, I would appreciate your input on this. I've seen that this was already discussed here, although in a different context: https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#r529172740

Reply:
For the case where `--keycloak-groups` & `--keycloak-roles` is set, if you want to enforce that a user has at least 1 group AND at least 1 role that matches, you should override the default `Authorize` method in the Keycloak provider.

I'd imagine you'd manage 2 sets (groups and roles) and force membership checks in BOTH sets to return true to result in `Authorize` returning true.

Reply:
I think that would make more sense, what do you think? Should I go ahead and do that then?

Reply:
Yeah, if you need separate ANY logic between groups and roles, and between the sets AND logic, overriding the default with a custom `Authorize` method that does that logic is best.